cmd/wstail/forwarders.go

```diff
@@ -1,106 +1,107 @@
 
+	"github.com/ghodss/yaml"
 	"github.com/gravitational/trace"
 	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
 )
 
-// updateForwarders updates log forwarder configuration and reloads the logging service.
-// It receives new forwarder configuration, updates the configuration and restarts the rsyslog daemon
-// to force it to reload the configuration
-func updateForwarders(w http.ResponseWriter, r *http.Request) (err error) {
-	if r.Method != "PUT" {
-		return trace.BadParameter("invalid HTTP method: %v", r.Method)
+// initLogForwarders reads config map that contains log forwarder resources and creates
+// respective rsyslog configuration files
+func initLogForwarders() error {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return trace.Wrap(err)
 	}
-	var forwarders []forwarders.Forwarder
-	if err = readJSON(r, &forwarders); err != nil {
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	// Remove previous forwarder configuration files
-	if err = filepath.Walk(rsyslogConfigDir, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if path == rsyslogConfigDir {
-			return nil
-		}
-		if info.IsDir() {
-			return filepath.SkipDir
-		}
-		log.Infof("removing %v", path)
-		return trace.Wrap(os.Remove(path), "failed to remove %v", path)
-	}); err != nil {
-		log.Warningf("failed to delete forwarder configuration files: %v", err)
+	configMap, err := client.ConfigMaps(systemNamespace).Get(
+		forwardersConfigMap, metav1.GetOptions{})
+	if err != nil {
+		return trace.Wrap(err)
 	}
 
-	// Write new forwarder configuration
-	for _, forwarder := range forwarders {
-		path := path(forwarder)
-		f, err := os.Create(path)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		_, err = f.WriteString(config(forwarder))
-		if errClose := f.Close(); errClose != nil {
-			log.Warningf("failed to close file %v: %v", path, errClose)
-		}
+	if len(configMap.Data) == 0 {
+		log.Infof("no log forwarders configured")
+		return nil
+	}
+
+	for _, data := range configMap.Data {
+		err := initLogForwarder([]byte(data))
 		if err != nil {
 			return trace.Wrap(err)
 		}
 	}
 
-	// Reload rsyslogd
-	if out, err := exec.Command(rsyslogInitScript, "restart").CombinedOutput(); err != nil {
-		return trace.Wrap(err, "failed to restart rsyslogd: %s", out)
-	}
-
-	log.Infof("forwarder configuration updated")
 	return nil
 }
 
-const rsyslogConfigDir = "/etc/rsyslog.d"
+// initLogForwarders configures a single log forwarder from data found in config map
+func initLogForwarder(data []byte) error {
+	var forwarder logForwarder
 
-const rsyslogInitScript = "/etc/init.d/rsyslog"
+	err := yaml.Unmarshal([]byte(data), &forwarder)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = ioutil.WriteFile(
+		forwarderFilename(forwarder),
+		forwarderConfig(forwarder),
+		0755)
```

Comment:
`sharedExecutableMask`? Does it need to be executable? `sharedReadWriteMask` (`0666`)?